src/libmongoc/tests/test-mongoc-client-side-encryption.c

```diff
@@ -4181,6 +4181,387 @@ test_unique_index_on_keyaltnames (void *unused)
+      NULL /* reply */,
+      &error);
+   ASSERT_OR_PRINT (got, error);
+
+   cursor = mongoc_collection_aggregate (def->encryptedColl,
+                                         MONGOC_QUERY_NONE,
+                                         tmp_bson ("{}"),
+                                         NULL /* opts */,
+                                         NULL /* read prefs */);
+
+   got = mongoc_cursor_next (cursor, &found);
+   ASSERT_WITH_MSG (!got,
+                    "Expected error in mongoc_cursor_next, but got success");
+
+   ASSERT (mongoc_cursor_error (cursor, &error));
+
+   ASSERT_ERROR_CONTAINS (error, MONGOC_ERROR_SERVER, 123, "failpoint");
+
+   ASSERT_CMPSTR (def->aggEvent.gotType, "failed");
+   ASSERT_ERROR_CONTAINS (
+      def->aggEvent.gotFailedError, MONGOC_ERROR_SERVER, 123, "failpoint");
+   ASSERT (!got);
+   mongoc_cursor_destroy (cursor);
+   decryption_events_fixture_destroy (def);
+}
+
+/* test_decryption_events_network_error is a regression test for CDRIVER-4401.
+ * Send a command on an encrypted client resulting in a network error.
+ * Expect an error returned and a CommandFailed event to be emitted. */
+static void
+test_decryption_events_network_error (void *unused)
+
+{
+   bool got;
+   bson_error_t error;
+   decryption_events_fixture *def = decryption_events_setup ();
+   const bson_t *found;
+   mongoc_cursor_t *cursor;
+
+   got = mongoc_client_command_simple (
+      def->setupClient,
+      "admin",
+      tmp_bson ("{'configureFailPoint': 'failCommand', 'mode': {'times': 1}, "
+                "'data': {'errorCode': 123, 'closeConnection': true, "
+                "'failCommands': ['aggregate']}}"),
+      NULL /* read prefs */,
+      NULL /* reply */,
+      &error);
+   ASSERT_OR_PRINT (got, error);
+
+   cursor = mongoc_collection_aggregate (def->encryptedColl,
+                                         MONGOC_QUERY_NONE,
+                                         tmp_bson ("{}"),
+                                         NULL /* opts */,
+                                         NULL /* read prefs */);
+   got = mongoc_cursor_next (cursor, &found);
+   ASSERT_WITH_MSG (!got,
+                    "Expected error in mongoc_cursor_next, but got success");
+   ASSERT (mongoc_cursor_error (cursor, &error));
+   ASSERT_ERROR_CONTAINS (
+      error, MONGOC_ERROR_STREAM, MONGOC_ERROR_STREAM_SOCKET, "socket error");
+
+   ASSERT_CMPSTR (def->aggEvent.gotType, "failed");
+   ASSERT_ERROR_CONTAINS (def->aggEvent.gotFailedError,
+                          MONGOC_ERROR_STREAM,
+                          MONGOC_ERROR_STREAM_SOCKET,
+                          "socket error");
+   ASSERT (!got);
+   mongoc_cursor_destroy (cursor);
+   decryption_events_fixture_destroy (def);
+}
+
+/* test_decryption_events_decrypt_error is a regression test for CDRIVER-4401.
+ * Decrypt a reply with a malformed ciphertext.
+ * Expect an error returned and a CommandSucceeded event to be emitted with
+ * ciphertext. */
```

Comment:
CommandSucceeded is expected because the command itself succeeds, and only decryption fails, correct?

Reply:
Correct.